### PR TITLE
Several bug fixes

### PR DESCRIPTION
**Pull request by AlexanderSher**
- Fix issue #95: `from ... import` shows incorrect completions. `TryGetCompletionsInFromImport` didn't use information from `ModuleInfo._referencedModules`, trying to find module directly inside AST.
- Fix issue #109: Reloading modules leaks memory. `ModuleInfo.Clear` didn't remove back references from `ModuleReference` instances
- Fix issue #278: Huge memory usage when analysis 9 lines python code. Issue caused by Json.Net serializing a cycle in `DocumentSymbol` tree. `Server.ToDocumentSymbols` uses `Dictionary<IMemberResult, List<IMemberResult>>` to make a tree, and `MemberResult`'s `GetHashCode` and `Equals` didn't take care of the scope. As a result, all results with equal parent name were collapsed into one level, which created a loop in a resulting tree where Json.Net has been hanging.
- Some code clean-up (dead code paths, auto-properties, pattern matching, etc.)

### CODE REVIEW

**Comment by jakebailey** [comment]
I can confirm that this fixes #278. Nice!

File: src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs [hunk 1]
```diff
@@ -22,16 +22,14 @@ namespace Microsoft.PythonTools.Parsing.Ast {
     public class DottedName : Node {
         private readonly NameExpression[] _names;
 
-        public DottedName(NameExpression[] names) {
+        public DottedName(NameExpression[]/*!*/ names) {
```
Comment (jakebailey) [inline src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs]: What's the meaning of `/*!*/` introduced in a few of these files?

File: src/Analysis/Engine/Impl/Values/ModuleInfo.cs [hunk 2]
```diff
@@ -172,22 +161,27 @@ internal class ModuleInfo : AnalysisValue, IReferenceableContainer, IModuleInfo
...
+        public void AddModuleReference(string moduleName, ModuleReference moduleReference) {
+            Check.ArgumentNotNull(nameof(moduleName), moduleName);
+            Check.ArgumentNotNull(nameof(moduleReference), moduleReference);
+
+            _referencedModules[moduleName] = moduleReference;
+            moduleReference.AddReference(this);
+        }
+
+        public void ClearReferencedModules() {
+            foreach (var moduleReference in _referencedModules.Values.Distinct()) {
```
Comment (MikhailArkhipov) [inline src/Analysis/Engine/Impl/Values/ModuleInfo.cs]: Not sure if Values can be null (when module is missing) - perhaps `MaybeEnumerate`?

File: src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs (hunk 1 above)
Comment (AlexanderSher) [reply src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs]: Just a comment that I've used to mark parameter that can't be null. I can add `Check` calls instead.

File: src/Analysis/Engine/Impl/Values/ModuleInfo.cs (hunk 2 above)
Comment (AlexanderSher) [reply src/Analysis/Engine/Impl/Values/ModuleInfo.cs]: `_referencedModules` is a dictionary, it is always initialized and it is readonly.

File: src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs (hunk 1 above)
Comment (jakebailey) [reply src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs]: You don't have to change it, I just hadn't seen that before and was curious. Grepping through the project shows its use a number of other times as well (so I just hadn't noticed).

Comment (AlexanderSher) [reply src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs]: To be clear - this is just a comment. It isn't handled by any analyzers. I really expect to see non-nullable reference types in C# 8: https://github.com/dotnet/csharplang/wiki/Nullable-Reference-Types-Preview

Comment (jakebailey) [reply src/Analysis/Engine/Impl/Parsing/Ast/DottedName.cs]: Yep, I know. I actually went searching to see if that feature existed yet when you mentioned it. 🙂

**Pull request merged**